File: thih/erlang/unify.erl
% 6 Unification and Matching
-module(unify).
-export([mgu/2,varBind/2,match/2]).

mgu({tap,L,R}, {tap,L_,R_}) ->
  S1 = mgu(L, L_),
  S2 = mgu(subst:typeApply(S1, R), subst:typeApply(S1, R_)),
  subst:atat(S2, S1);
mgu({tvar,U}, Type) -> varBind(U, Type);
mgu(Type, {tvar,U}) -> varBind(U, Type);
mgu({tcon,Tc1}, {tcon,Tc2}) when Tc1 == Tc2 -> subst:nullSubst();
mgu(_, _) -> throw("types do not unify").

varBind(Tyvar, Type) ->
  Subst = if
    Type == {tvar,Tyvar} -> {ok,subst:nullSubst()};
    true -> {ng}
  end,

  Occurs = lists:member(Tyvar, type:typeTv(Type)),
  if Occurs -> throw("occurs check fails") end,
  KindCheck = (type:tyvarKind(Tyvar) /= type:typeKind(Type)),
  if KindCheck -> throw("kinds do not match") end,
  case Subst of
  	{ok,V} -> V;
  	{ng} -> type:tyvar_append(Tyvar, Type)
  end
  .

match({tap, L, R}, {tap, L_, R_}) ->
  Sl = match(L,L_),
  Sr = match(R,R_),
  subst:merge(Sl, Sr);
match({tvar,U}, T) ->
  Ck = (type:tyvarKind(U) == type:typeKind(T)),
  if
  	Ck -> type:tyvar_append(U,T);
  	true -> throw("types do not match")
  end;
match({tcon,Tc1}, {tcon,Tc2}) when(Tc1 == Tc2) -> subst:nullSubst();
match(_,_) -> throw("types do not match").

File: thih/ocaml2/unify.ml
(*|
# 6 Unification and Matching

    >>> open Unify ;;

*)

open List
open Kind
open Type
open Subst_

let rec mgu (t1:type_) (t2:type_):subst =
  match t1, t2 with
  | TAp(l, r), TAp(l', r') ->
    let s1 = mgu l l' in
    let s2 = mgu (typeApply s1 r) (typeApply s1 r') in
    s2 @@ s1
  | TVar u, t | t, TVar u -> varBind u t
  | TCon tc1, TCon tc2 when tc1 = tc2 -> nullSubst
  | _ -> failwith "types do not unify"

and varBind (u:tyvar) (t:type_):subst =
  match t with
  | _ when t = TVar u                -> nullSubst
  | _ when mem u (typeTv t)          -> failwith "occurs check fails"
  | _ when tyvarKind u <> typeKind t -> failwith "kinds do not match"
  | _                                -> u +-> t

let rec match_ (t1:type_) (t2:type_):subst =
  match t1, t2 with
  | TAp(l, r), TAp(l', r') ->
    let sl = match_ l l' in
    let sr = match_ r r' in
    merge sl sr
  | TVar u, t when tyvarKind u = typeKind t -> u +-> t
  | TCon tc1, TCon tc2 when tc1 = tc2 -> nullSubst
  | _ -> failwith "types do not match"

(*|
    
    >>> let t1 = TVar(Tyvar("a", Star)) ;;
    val t1 : Type.type_ = TVar (Tyvar ("a", Star))

    >>> let t2 = TVar(Tyvar("b", Star)) ;;
    val t2 : Type.type_ = TVar (Tyvar ("b", Star))

    >>> let tv1 = Tyvar("a", Star) ;;
    val tv1 : Type.tyvar = Tyvar ("a", Star)

    >>> let t3 = tInt ;;
    val t3 : Type.type_ = TCon (Tycon ("Int", Star))

## mgu

    >>> let subst = mgu(t1)(t2);;
    val subst : Subst_.subst = [(Tyvar ("a", Star), TVar (Tyvar ("b", Star)))]

    >>> subst =
        [(Tyvar("a", Star), TVar(Tyvar("b", Star)))];;
    - : bool = true

    >>> let subst2 = mgu(t1)(t3);;
    val subst2 : Subst_.subst = [(Tyvar ("a", Star), TCon (Tycon ("Int", Star)))]

    >>> subst2 =
        [(Tyvar("a", Star), tInt)];;
    - : bool = true

## varBind

    >>> let subst = varBind(tv1)(t1);;
    val subst : Subst_.subst = []

    >>> subst = [];;
    - : bool = true

    >>> let subst2 = varBind(tv1)(t3);;
    val subst2 : Subst_.subst = [(Tyvar ("a", Star), TCon (Tycon ("Int", Star)))]

    >>> subst2 =
        [(Tyvar("a", Star), tInt)];;
    - : bool = true

## match_

    >>> let subst = match_(t1)(t2);;
    val subst : Subst_.subst = [(Tyvar ("a", Star), TVar (Tyvar ("b", Star)))]

    >>> subst =
        [(Tyvar("a", Star), TVar(Tyvar("b", Star)))];;
    - : bool = true

    >>> let subst2 = match_(t1)(t3);;
    val subst2 : Subst_.subst = [(Tyvar ("a", Star), TCon (Tycon ("Int", Star)))]

    >>> subst2 = [(Tyvar("a", Star), tInt)] ;;
    - : bool = true

*)
